src/Feature/Editor/Draw.re

```diff
@@ -0,0 +1,250 @@
+  let fontMetrics = Revery.Font.getMetrics(font, editorFont.fontSize);
+
+  {
+    canvasContext,
+    width,
+    height,
+    scrollX,
+    scrollY,
+    lineHeight,
+    font,
+    fontMetrics,
+    fontSize: editorFont.fontSize,
+    charWidth: editorFont.measuredWidth,
+    charHeight: editorFont.measuredHeight,
+  };
+};
+
+let renderImmediate = (~context, ~count, render) =>
+  ImmediateList.render(
+    ~scrollY=context.scrollY,
+    ~rowHeight=context.lineHeight,
+    ~height=float(context.height),
+    ~count,
+    ~render=(i, offsetY) => render(i, offsetY +. context.scrollY),
+    (),
+  );
+
+let drawRect = (~context, ~x, ~y, ~width, ~height, ~paint) =>
+  CanvasContext.drawRectLtwh(
+    ~left=x -. context.scrollX,
+    ~top=y -. context.scrollY,
+    ~width,
+    ~height,
+    ~paint,
+    context.canvasContext,
+  );
+let rect = drawRect;
+
+let drawText = (~context, ~x, ~y, ~paint, text) =>
+  CanvasContext.drawText(
+    ~x=x -. context.scrollX,
+    ~y=y -. context.scrollY,
+    ~paint,
+    ~text,
+    context.canvasContext,
+  );
+let text = drawText;
+
+let underline =
+    (~context, ~buffer, ~leftVisibleColumn, ~color=Colors.black, r: Range.t) => {
+  let line = Index.toZeroBased(r.start.line);
+  let start = Index.toZeroBased(r.start.column);
+  let endC = Index.toZeroBased(r.stop.column);
+
+  let text = Buffer.getLine(line, buffer);
+  let (startOffset, _) =
+    BufferViewTokenizer.getCharacterPositionAndWidth(
+      ~viewOffset=leftVisibleColumn,
+      text,
+      start,
+    );
+  let (endOffset, _) =
+    BufferViewTokenizer.getCharacterPositionAndWidth(
+      ~viewOffset=leftVisibleColumn,
+      text,
+      endC,
+    );
+
+  let paint = Skia.Paint.make();
+  Skia.Paint.setColor(paint, Revery.Color.toSkia(color));
+
+  drawRect(
+    ~context,
+    ~x=float(startOffset) *. context.charWidth,
+    ~y=
+      context.charHeight
+      *. float(Index.toZeroBased(r.start.line))
+      +. (context.charHeight -. 2.),
+    ~height=1.,
+    ~width=max(float(endOffset - startOffset), 1.0) *. context.charWidth,
+    ~paint,
+  );
+};
+
+let range =
+    (
+      ~context,
+      ~padding=0.,
+      ~buffer,
+      ~leftVisibleColumn,
+      ~color=Colors.black,
+      r: Range.t,
+    ) => {
+  let doublePadding = padding *. 2.;
+  let line = Index.toZeroBased(r.start.line);
+  let start = Index.toZeroBased(r.start.column);
+  let endC = Index.toZeroBased(r.stop.column);
+
+  let lines = Buffer.getNumberOfLines(buffer);
+  if (line < lines) {
+    let text = Buffer.getLine(line, buffer);
+    let (startOffset, _) =
+      BufferViewTokenizer.getCharacterPositionAndWidth(
+        ~viewOffset=leftVisibleColumn,
+        text,
+        start,
+      );
+    let (endOffset, _) =
+      BufferViewTokenizer.getCharacterPositionAndWidth(
+        ~viewOffset=leftVisibleColumn,
+        text,
+        endC,
+      );
+    let length = max(float(endOffset - startOffset), 1.0);
+
+    let paint = Skia.Paint.make();
```

Comment:
I believe this will regress some of the perf work we did for Skia. Allocations like this end up pretty expensive (they build up promoted memory and make more work for the GC).

To see the impact - we have a `drawText` benchmark in Revery:
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  BENCHMARK                                                     |  ITERATIONS |  TIME               |  MINOR GC  |  MAJOR GC  |  MINOR ALLOC  |  PROMOTED  |  MAJOR ALLOC  |
|----------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Draw: drawText                                                |  10000      |  0.110551118851     |  1         |  0         |  230784       |  25        |  25           |
|----------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
```

This is with a pre-allocated `Skia.Paint.make`, and then if we allocate one every time:
```
|----------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Draw: drawText                                                |  10000      |  0.118616104126     |  2         |  0         |  320790       |  74165     |  74165        |
|----------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|

```

The promoted memory goes from `25` -> `74165`, and it's about a 9% increase in time. It adds up in our 'hot paths'.

Definitely going to have to do another round of perf measurement / optimization after this. With Skia, I can at least put together an end-to-end benchmark which will help measure and guide optimizations.